cli/internal/fs/turbo_json.go

```diff
@@ -162,18 +163,33 @@ func (c *TaskDefinition) UnmarshalJSON(data []byte) error {
 	} else {
 		c.ShouldCache = *rawPipeline.Cache
 	}
-	c.EnvVarDependencies = []string{}
+
+	envVarDependencies := make(util.Set)
 	c.TopologicalDependencies = []string{}
 	c.TaskDependencies = []string{}
+
 	for _, dependency := range rawPipeline.DependsOn {
 		if strings.HasPrefix(dependency, envPipelineDelimiter) {
-			c.EnvVarDependencies = append(c.EnvVarDependencies, strings.TrimPrefix(dependency, envPipelineDelimiter))
+			envVarDependencies.Add(strings.TrimPrefix(dependency, envPipelineDelimiter))
 		} else if strings.HasPrefix(dependency, topologicalPipelineDelimiter) {
 			c.TopologicalDependencies = append(c.TopologicalDependencies, strings.TrimPrefix(dependency, topologicalPipelineDelimiter))
 		} else {
 			c.TaskDependencies = append(c.TaskDependencies, dependency)
 		}
 	}
+
+	// Append env key into EnvVarDependencies
+	for _, value := range rawPipeline.Env {
+		if strings.HasPrefix(value, envPipelineDelimiter) {
+			// Log a warning, because dependsOn accepts `$` prefixed strings to specify
+			// environment variable dependencies. Because we don't strip out the `$` character
+			// from the env key, it may be confusing to end users, who are just trying to move it over.
+			log.Printf("[WARNING] You specified \"%s\" in the \"env\" key. This is treated as a literal, you may want to remove the \"$\".", value)
```

Comment:
I wouldn't include this warning. In general, people ignore warnings unless something breaks. If something breaks, they'll investigate. When something does break, we don't know that it will be related to this, or any other warning that we might include. In the worst case, we send them on a false trail because whatever broke is unrelated. The best help we can give them is to show what commands were invoked with what inputs. This is where something like a run summary comes into play. 

For this warning in particular, users of this API will fall into two camps:
 1. Users coming from the previous behavior of `dependsOn`.
 2. Users new to `turbo` who have never seen the previous syntax

We expect going forward that group 2 will be the much larger group. They will have no reason to think they need a `$` prefix. In the event of an unlikely subset who do intend to set an environment variable that starts with a `$`, they will then have a warning they can't get rid of.

Group 1 will have a codemod to move them to the proper syntax.

Reply:
Also, if we do end up keeping this warning, we need to thread our `cli.Ui` instance in here, so that it can respect whatever output is configured. We shouldn't be using `log` directly.

Reply:
ok, that's fair too. I've removed the warning

Reply:
@gsoltis the issue here is that somebody will add "$FOO" to the env, get no feedback, and just assume that it works.

Later, when their cache isn't segmented, they'll have _no idea why_ and (at this time) have no way to inspect the build to understand what happened.

In other words, failing to use this API correctly _will_ ruin your day. We should aim to reduce the incidence of these types of failure modes which is why I propose automatically removing the `$`.

Reply:
Reached consensus in slack to throw a hard error for `$` prefixed strings